programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_transaction.rs

```diff
@@ -39,7 +40,9 @@ pub fn process_insert_two_leaves<'info, 'a>(
         };
 
         // Insert the pair into the merkle tree
-        merkle_tree.merkle_tree.insert(*leaf_left, *leaf_right)?;
+        // TODO(vadorovsky): Support appending two leaves in concurrent MT.
+        let leaf = Poseidon::hashv(&[leaf_left.as_slice(), leaf_right.as_slice()])?;
+        merkle_tree.merkle_tree.append(&leaf)?;
```

Comment:
if you do this you will probably need to decrease the mt height by one to 21 right?

Reply:
tbh I my intuition is that it's cleaner to put it into the crate right away.

Reply:
Yeah, good idea, I will just add it to the crate as `append_two` method.

Reply:
Or maybe even something like

```rust
pub fn append_many<const N: usize>(leaves: &[[u8;32]; N]) {
        let leaf = Poseidon::hashv(leaves)?;
        merkle_tree.merkle_tree.append(&leaf)?;
}
```

Would that make sense? I remember you mentioning that we would want to support more than 2 leaves at some point.

Reply:
if we do this we don't have a binary tree at the leaf level anymore

Reply:
but you are right we can just as well add a batch update instruction which works for any number, but let's do that in a separate pr because this pr is blocking for public transactions

Reply:
this is the key part which resets the algorithm to start from zero again https://github.com/Lightprotocol/light-protocol/blob/1834b3ef0202b9f1aa4934982c8414f1af6119b6/programs/merkle-tree/src/transaction_merkle_tree/update_merkle_tree_lib/instructions.rs#L60-L69

Reply:
- the idea is update subtrees until you hit a zero value
- once you hit a zero value reset the algorithm and start inserting the next leaf, you can do that with any number of leaves

Reply:
I think it needs some tweaking to get it working in concurrent Merkle trees - we are not operating on subtrees (like Tornado), but on changelog (full Merkle proofs). Something to keep in mind when adjusting your algorithm.

Reply:
We decided to just do 2 updates for now and worry about the algo tweak in a separate PR.